File: app/src/main/java/com/example/firefighters/viewmodels/EmergencyViewModel.java
package com.example.firefighters.viewmodels;

import com.example.firefighters.models.EmergencyModel;
import com.example.firefighters.repositories.EmergencyRepository;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

public class EmergencyViewModel extends ViewModel {

    private EmergencyRepository repository;

    public void init() {
        if (repository != null)
            return;
        repository = EmergencyRepository.getInstance();
    }
    public LiveData<QuerySnapshot> getEmergenciesQuery(DocumentSnapshot lastDocument, String lastFilter, Query.Direction lastOrder, int limitCount) {
        return repository.getEmergenciesQuery(lastDocument, lastFilter, lastOrder, limitCount);
    }
    public LiveData<QuerySnapshot> getEmergenciesQuerySnapshot(String lastFilter, Query.Direction lastOrder) {
        return repository.getEmergenciesQuerySnapshot(lastFilter, lastOrder);
    }

    public LiveData<EmergencyModel> getEmergencyWorkingOnModel(String  currentUnit) {
        return repository.getEmergencyWorkingOnModel(currentUnit);
    }
    public LiveData<QuerySnapshot> getEmergencyWorkingOn(String  currentUnit) {
        return repository.getEmergencyWorkingOn(currentUnit);
    }
    public LiveData<Integer> saveEmergency(EmergencyModel emergencyModel) {
        return repository.saveEmergency(emergencyModel);
    }
    public LiveData<Integer> deleteEmergency(EmergencyModel emergencyModel) {
        return repository.deleteEmergency(emergencyModel);
    }
    public LiveData<Integer> updateEmergency(EmergencyModel emergencyModel) {
        return repository.updateEmergency(emergencyModel);
    }

}

File: app/src/main/java/com/example/firefighters/viewmodels/models/EmergencyViewModel.kt
package com.example.firefighters.viewmodels.models

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.firefighters.models.EmergencyModel
import com.example.firefighters.repositories.EmergencyRepository
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.QuerySnapshot

class EmergencyViewModel(private val repository: EmergencyRepository) : ViewModel() {

    fun getEmergenciesQuery(
        lastDocument: DocumentSnapshot?,
        lastFilter: String?,
        lastOrder: Query.Direction?,
        limitCount: Int
    ): LiveData<QuerySnapshot> {
        return repository.getEmergenciesQuery(lastDocument, lastFilter, lastOrder, limitCount)
    }

    fun getEmergenciesQuerySnapshot(
        lastFilter: String?,
        lastOrder: Query.Direction?
    ): LiveData<QuerySnapshot?> {
        return repository.getEmergenciesQuerySnapshot(lastFilter, lastOrder)
    }

    fun getLastQuerySnapshot(
        lastFilter: String?,
        lastOrder: Query.Direction?
    ): LiveData<QuerySnapshot?> {
        return repository.getLastQuerySnapshot(lastFilter, lastOrder)
    }

    fun getEmergencyWorkingOnModel(currentUnit: String?): LiveData<EmergencyModel?> {
        return repository.getEmergencyWorkingOnModel(currentUnit)
    }

    fun getEmergencyWorkingOn(currentUnit: String?): LiveData<QuerySnapshot?> {
        return repository.getEmergencyWorkingOn(currentUnit)
    }

    fun saveEmergency(emergencyModel: EmergencyModel?): LiveData<Int> {
        return repository.saveEmergency(emergencyModel!!)
    }

    fun deleteEmergency(emergencyModel: EmergencyModel?): LiveData<Int> {
        return repository.deleteEmergency(emergencyModel!!)
    }

    fun updateEmergency(emergencyModel: EmergencyModel?): LiveData<Int> {
        return repository.updateEmergency(emergencyModel!!)
    }

    class Factory(
        private val repository: EmergencyRepository
    ) : ViewModelProvider.NewInstanceFactory() {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return EmergencyViewModel(repository) as T
        }
    }
}